src/engine/LocalVocab.h

```diff
@@ -50,60 +54,68 @@ class LocalVocab {
-  // the function runs in linear time in the number of word sets.
+  // Make a logical copy, where all sets of `LocalVocabEntry`s become static
+  // and the primary set becomes empty. This only copies shared pointers and
+  // takes time linear in the number of sets.
   LocalVocab clone() const;
 
   // Moving a local vocabulary is not problematic (though the typical use case
-  // in our code is to copy shared pointers to local vocabularies).
+  // in our code is to copy shared pointers from one `LocalVocab` to another).
   LocalVocab(LocalVocab&&) = default;
   LocalVocab& operator=(LocalVocab&&) = default;
 
-  // Get the index of a word in the local vocabulary. If the word was already
-  // contained, return the already existing index. If the word was not yet
-  // contained, add it, and return the new index.
-  LocalVocabIndex getIndexAndAddIfNotContained(const LiteralOrIri& word);
-  LocalVocabIndex getIndexAndAddIfNotContained(LiteralOrIri&& word);
+  // For a given `LocalVocabEntry`, return the corresponding `LocalVocabIndex`
+  // (which is just the address of the `LocalVocabEntry`). If the
+  // `LocalVocabEntry` is not contained in any of the sets, add it to the
+  // primary.
+  LocalVocabIndex getIndexAndAddIfNotContained(const LocalVocabEntry& word);
+  LocalVocabIndex getIndexAndAddIfNotContained(LocalVocabEntry&& word);
 
-  // Get the index of a word in the local vocabulary, or std::nullopt if it is
-  // not contained. This is useful for testing.
+  // Like `getIndexAndAddIfNotContained`, but if the `LocalVocabEntry` is not
+  // contained in any of the sets, do not add it and return `std::nullopt`.
   std::optional<LocalVocabIndex> getIndexOrNullopt(
-      const LiteralOrIri& word) const;
+      const LocalVocabEntry& word) const;
 
-  // The number of words in the vocabulary.
-  // Note: This is not constant time, but linear in the number of word sets.
+  // The number of words in this local vocabulary.
   size_t size() const {
-    auto result = primaryWordSet().size();
-    for (const auto& previous : otherWordSets_) {
-      result += previous->size();
+    if constexpr (ad_utility::areExpensiveChecksEnabled) {
+      auto size = primaryWordSet().size();
+      for (const auto& previous : otherWordSets_) {
+        size += previous->size();
+      }
+      AD_CORRECTNESS_CHECK(size == size_);
     }
-    return result;
+    return size_;
   }
 
   // Return true if and only if the local vocabulary is empty.
   bool empty() const { return size() == 0; }
 
-  // Return a const reference to the word.
-  const LiteralOrIri& getWord(LocalVocabIndex localVocabIndex) const;
-
-  // Create a local vocab that contains and keeps alive all the words from each
-  // of the `vocabs`. The primary word set of the newly created vocab is empty.
-  static LocalVocab merge(std::span<const LocalVocab*> vocabs);
+  // Get the `LocalVocabEntry` corresponding to the given `LocalVocabIndex`.
```

Comment:
This function is only there for legacy reason,s and could be thrown out in the future, as we can simply dereference the pointers, add a NOTE here.